### PR TITLE
Skip dry run on node upgrade.

### PR DESCRIPTION
**Pull request by neykov**
Add parameter to choose whether to launch a dry run node or to upgrade directly.
Allow extending classes to customize the effector parameters by overriding a method.

Depends on https://github.com/apache/incubator-brooklyn/pull/413

### CODE REVIEW

**Comment by neykov** [comment]
Failure looks unrelated, local tests pass. Rebased to force a rebuild.

**Comment by neykov** [comment]
Forcing another update, local tests passing again - what's with the failures...

**Comment by neykov** [comment]
Failure could be caused by https://github.com/apache/incubator-brooklyn/pull/413, investigating...

**Comment by neykov** [comment]
Test failure should be fixed in https://github.com/apache/incubator-brooklyn/pull/482 (not included in this PR). Rebasing again to kick another build.

File: software/base/src/main/java/brooklyn/entity/brooklynnode/effector/BrooklynNodeUpgradeEffectorBody.java [hunk 1]
```diff
@@ -132,44 +193,34 @@ public Void call(ConfigBag parametersO) {
...
+    }
 
-        DynamicTasks.waitForLast();
-        Entities.unmanage(dryRunChild);
+    protected EntitySpec<? extends BrooklynNode> createDryRunSpec() {
+        return EntitySpec.create(BrooklynNode.class);
+    }
 
-        return null;
+    protected SshMachineLocation getSshMachineLocation() {
```
Comment (ahgittin) [inline software/base/src/main/java/brooklyn/entity/brooklynnode/effector/BrooklynNodeUpgradeEffectorBody.java]: remove this method; you can use `Locations.findUniqueSshMachineLocation(entity().getLocations()).get()`

File: software/base/src/main/java/brooklyn/entity/brooklynnode/effector/BrooklynNodeUpgradeEffectorBody.java [hunk 2]
```diff
@@ -132,44 +193,34 @@ public Void call(ConfigBag parametersO) {
...
-                String runDir = entity().getAttribute(SoftwareProcess.RUN_DIR);
-                String bkDir = Urls.mergePaths(runDir, "..", Urls.getBasename(runDir)+"-backups", dryRunNodeUid);
-                String dryRunDir = Preconditions.checkNotNull(dryRunChild.getAttribute(SoftwareProcess.RUN_DIR));
-                log.debug(this+" storing backup of previous version in "+bkDir);
-                DynamicTasks.queue(SshEffectorTasks.ssh(
-                    "cd "+runDir,
-                    "mkdir -p "+bkDir,
-                    "mv * "+bkDir,
-                    "cd "+dryRunDir,
-                    "mv * "+runDir
```
Comment (ahgittin) [inline software/base/src/main/java/brooklyn/entity/brooklynnode/effector/BrooklynNodeUpgradeEffectorBody.java]: looks like we lose the two lines above?  i think that's fine, but wanted to check.

it means the `runDir` contents will have to be repopulated from the (already downloaded) `installDir`, costing perhaps a few extra seconds of downtime.  should be easy enough to add these two statements back in the case where dryRun has just been done if we wanted to.  but it does feel a bit dodgy.  no strong feelings...

**Comment by ahgittin** [comment]
LGTM, minor comments, then can merge

File: software/base/src/main/java/brooklyn/entity/brooklynnode/effector/BrooklynNodeUpgradeEffectorBody.java (hunk 1 above)
Comment (neykov) [inline software/base/src/main/java/brooklyn/entity/brooklynnode/effector/BrooklynNodeUpgradeEffectorBody.java]: Removed it altogether, couldn't find any usages (even in subclasses).

File: software/base/src/main/java/brooklyn/entity/brooklynnode/effector/BrooklynNodeUpgradeEffectorBody.java (hunk 2 above)
Comment (neykov) [inline software/base/src/main/java/brooklyn/entity/brooklynnode/effector/BrooklynNodeUpgradeEffectorBody.java]: If I remember correctly, this was done on purpose to force the customize step and regenerate the local-brooklyn.properties. Definitely worth a comment.

**Comment by neykov** [comment]
Addressed comments.

**Comment by ahgittin** [comment]
i did wonder whether that method was used at all!  makes sense about the customize step.  merging.

**Pull request merged**